File: apps/ios/UNESKit/Sources/UNESKit/Features/Paradoxo/ParadoxoFormat.swift
import Foundation

enum ParadoxoFormat {
    /// Grouped integer — "6.868" (pt-BR) / "6,868" (en).
    static func count(_ value: Int, locale: Locale = .autoupdatingCurrent) -> String {
        value.formatted(.number.grouping(.automatic).locale(locale))
    }

    /// Whole percent — "86%".
    static func percent(_ value: Int, locale: Locale = .autoupdatingCurrent) -> String {
        (Double(value) / 100).formatted(.percent.precision(.fractionLength(0)).locale(locale))
    }

    /// Signed one-decimal delta — "+1,8" / "−2,3" (true minus sign).
    static func signedGrade(_ value: Double, locale: Locale = .autoupdatingCurrent) -> String {
        let sign = value < 0 ? "−" : "+"
        return sign + formatGrade(abs(value), locale: locale)
    }

    /// The hero headline: trend deltas are signed, everything else reads as
    /// a grade.
    static func metric(of fact: ParadoxoPulseFact, locale: Locale = .autoupdatingCurrent) -> String {
        switch fact.kind {
        case .trend, .rising:
            signedGrade(fact.metric, locale: locale)
        case .brutal, .kind, .gap, .surprise, .signature:
            formatGrade(fact.metric, locale: locale)
        }
    }
}

File: apps/android/app/src/main/kotlin/dev/forcetower/unes/ui/feature/paradoxo/ParadoxoFormat.kt
package dev.forcetower.unes.ui.feature.paradoxo

import dev.forcetower.melon.feature.paradoxo.domain.model.ParadoxoPulseFact
import dev.forcetower.melon.feature.paradoxo.domain.model.ParadoxoPulseKind
import java.text.NumberFormat
import kotlin.math.abs
import kotlin.math.floor

// Number formatting for the aggregates. Grades truncate to one decimal
// (6,95 → 6,9 — never round toward approval, mirrors iOS `formatGrade`);
// counts group by the device locale ("6.868" in pt-BR).
internal object ParadoxoFormat {

    fun grade(value: Double?): String {
        if (value == null || value.isNaN()) return "—"
        val truncated = floor(value * 10) / 10.0
        return "%.1f".format(truncated).replace('.', ',')
    }

    // True minus sign, matching iOS `signedGrade` ("+1,8" / "−2,3").
    fun signedGrade(value: Double): String {
        val sign = if (value < 0) "−" else "+"
        return sign + grade(abs(value))
    }

    fun count(value: Int): String = NumberFormat.getIntegerInstance().format(value.toLong())

    fun percent(value: Int): String = "$value%"

    // Trend/rising pulse cards show a signed delta; every other kind shows a
    // plain grade.
    fun metric(fact: ParadoxoPulseFact): String = when (fact.kind) {
        ParadoxoPulseKind.Trend, ParadoxoPulseKind.Rising -> signedGrade(fact.metric)
        else -> grade(fact.metric)
    }
}
